src/autora/state/delta.py

```diff
@@ -74,25 +75,179 @@ class State:
+        CoerceStateList(o=None, p=['n', 'o', 't', ' ', 'a', ' ', 'l', 'i', 's', 't'])
+
+        If the input data are of the correct type, they are returned unaltered:
+        >>> r + Delta(p=["a", "list"])
+        CoerceStateList(o=None, p=['a', 'list'])
+
+        With a converter, inputs are converted to the type output by the converter:
+        >>> import pandas as pd
+        >>> @dataclass(frozen=True)
+        ... class CoerceStateDataFrame(State):
+        ...    q: pd.DataFrame = field(default_factory=pd.DataFrame,
+        ...                            metadata={"delta": "replace",
+        ...                                      "converter": pd.DataFrame})
+
+        If the type is already correct, the object is passed to the converter,
+        but should be returned unchanged:
+        >>> s = CoerceStateDataFrame()
+        >>> (s + Delta(q=pd.DataFrame([("a",1,"alpha"), ("b",2,"beta")], columns=list("xyz")))).q
+           x  y      z
+        0  a  1  alpha
+        1  b  2   beta
+
+        If the type is not correct, the object is converted if possible. For a dataframe,
+        we can convert records:
+        >>> (s + Delta(q=[("a",1,"alpha"), ("b",2,"beta")])).q
+           0  1      2
+        0  a  1  alpha
+        1  b  2   beta
+
+        ... or an array:
+        >>> (s + Delta(q=np.linspace([1, 2], [10, 15], 3))).q
+              0     1
+        0   1.0   2.0
+        1   5.5   8.5
+        2  10.0  15.0
+
+        ... or a dictionary:
+        >>> (s + Delta(q={"a": [1,2,3], "b": [4,5,6]})).q
+           a  b
+        0  1  4
+        1  2  5
+        2  3  6
+
+        ... or a list:
+        >>> (s + Delta(q=[11, 12, 13])).q
+            0
+        0  11
+        1  12
+        2  13
+
+        ... but not, for instance, a string:
+        >>> (s + Delta(q="not compatible with pd.DataFrame")).q
+        Traceback (most recent call last):
+        ...
+        ValueError: DataFrame constructor not properly called!
+
+        Without a converter:
+        >>> @dataclass(frozen=True)
+        ... class CoerceStateDataFrameNoConverter(State):
+        ...    r: pd.DataFrame = field(default_factory=pd.DataFrame, metadata={"delta": "replace"})
+
+        ... there is no coercion – the object is passed unchanged
+        >>> t = CoerceStateDataFrameNoConverter()
+        >>> (t + Delta(r=np.linspace([1, 2], [10, 15], 3))).r
+        array([[ 1. ,  2. ],
+               [ 5.5,  8.5],
+               [10. , 15. ]])
+
+
+        A converter can cast from a DataFrame to a np.ndarray (with a single datatype),
+        for instance:
+        >>> import numpy as np
+        >>> @dataclass(frozen=True)
+        ... class CoerceStateArray(State):
+        ...    r: Optional[np.ndarray] = field(default=None,
+        ...                            metadata={"delta": "replace",
+        ...                                      "converter": np.asarray})
+
+        Here we pass a dataframe, but expect a numpy array:
```

Comment:
Quick question: Do we also allow for casting from pandas to numpy? I think we should allow that since most people will be feeding the theorists with numpy array (it's just simpler).

Reply:
You can always put a `pd.DataFrame` through `np.array` to get a numpy array. 

My recommendation would be to have the functions and classes which want to represent the data internally as numpy arrays accept `np.typing.ArrayLike` as the input type, which allows for lots of input types – lists of values, or `DataFrames` or `np.ndarrays` – and then do a `np.asarray()` call at the top of the function to turn the input into the array you want. 

It's really hard to do the casting properly from outside these functions, as we'd have to inspect the function signatures and work out whether the type we have is compatible. It's much simpler and more efficient to do this within the function itself, especially if we have a general interchange format like pd.DataFrame as our standard.